### PR TITLE
tour of beam integration tests

### LINKED ISSUE #24982: [Task][Tour of Beam][Frontend]: Add unit and integration tests to Tour of Beam frontend.
### What needs to happen?

Add unit and integration tests to the Tour of Beam frontend app.

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by Malarg**
tour of beam integration tests
resolves #24982

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: .github/workflows/tour_of_beam_frontend_test.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+name: Tour Of Beam Fronend Test
+
+on:
+  push:
+    paths: ['learning/tour-of-beam/frontend/**', 'playground/frontend/playground_components']
+    branches: ['**']
+  pull_request:
+    paths: ['learning/tour-of-beam/frontend/**', 'playground/frontend/playground_components']
+    branches: ['**']
+  workflow_dispatch:
```
Comment (alexeyinkin) [inline .github/workflows/tour_of_beam_frontend_test.yml:25]: Delete.

File: playground/frontend/playground_components_dev/lib/src/common_tests/toggle_brightness_mode_test.dart [hunk 2]
```diff
@@ -18,7 +18,8 @@
 
 import 'package:flutter/material.dart';
 import 'package:flutter_test/flutter_test.dart';
-import 'package:playground_components_dev/playground_components_dev.dart';
+
+import '../../playground_components_dev.dart';
```
Comment (alexeyinkin) [inline playground/frontend/playground_components_dev/lib/src/common_tests/toggle_brightness_mode_test.dart:22]: Please import specific files.

File: .github/workflows/tour_of_beam_frontend_test.yml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      # - name: 'Test tour of beam'
+      #   working-directory: learning/tour-of-beam/frontend
+      #   run: flutter test
+
+      - uses: nanasess/setup-chromedriver@v1
+
+      - name: 'Integration tests'
+        run: |
+          chromedriver --port=4444 &
+          ./gradlew :learning:tour-of-beam:frontend:integrationTest -PdeviceId=web-server
```
Comment (damondouglas) [inline .github/workflows/tour_of_beam_frontend_test.yml:84]: This needs to go into the tob frontend README.

**Review by damondouglas — CHANGES_REQUESTED** [review]
1. Running the application, I received a `bottom overflowed`  error.  Is it possible to perform resizing in integration tests and detect bottom overflowed error when resizing, etc.?

image

2. After running `./gradlew :learning:tour-of-beam:frontend:integrationTest -PdeviceId=web-server` I saw that the a `learning/tour-of-beam/frontend/flutter_01.log` was added to untracked files and `learning/tour-of-beam/frontend/pubspec.lock` was modified.

3. Follow comments about the misalignment of the tob frontend README and the GitHub actions yaml. I had to follow yaml to get this to work and spent a lot of time figuring out what was wrong due to following the README.

**Comment by Malarg** [comment]
@damondouglas bottom overflowed error will be fixed here: https://github.com/apache/beam/pull/25964
Also I didn't got flutter_01.log file after executing integration tests. If the file will be generated at your machine again, please attach it.

**Review by damondouglas — CHANGES_REQUESTED** [review]
Shouldn't the integration tests detect "bottom overflowed" which may be one of the most common flutter errors?  I saw a note that this will be address in a future issue but shouldn't the integration tests at least detect this and have this test skipped until corrected?

**Comment by alexeyinkin** [comment]
Run Java PreCommit
Run Java_Examples_Dataflow PreCommit
Run Java_IOs_Direct PreCommit
Run Python_Integration PreCommit

**Review by damccorm — COMMENTED** [review]
Mostly LGTM, just had one question and one nit

File: .github/workflows/playground_frontend_test.yml [hunk 4]
```diff
@@ -22,7 +22,6 @@ on:
   pull_request:
     paths: ['playground/frontend/**']
     branches: ['**']
-  workflow_dispatch:
```
[the remark this answers is not in the document]
Comment (alexeyinkin) [reply .github/workflows/playground_frontend_test.yml:25]: That was accidental. Re-enabled, added for Tour of Beam as well.

**Comment by damccorm** [comment]
Run Python_Integration PreCommit

**Review by damccorm — APPROVED** [review]
Thanks!

**Pull request merged**